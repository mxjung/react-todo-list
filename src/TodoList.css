.todoApp {
  width: 600px;
  display: flex;
  flex-direction: column;
  background-color: azure;
  align-items: center;

}

.todolist {
  height: 400px;
  width: 400px;
  background-color: beige;
  position: relative;
  display: flex;
  flex-direction: column;
  margin: 30px;
}
