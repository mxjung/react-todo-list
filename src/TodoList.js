import React, { useState } from "react";
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import {v4 as uuid} from "uuid";
import './TodoList.css';

/**
 *  TodoList - this component should render the NewTodoForm component and should 
 *  render the list of Todo components. Place your state that contains all of the 
 *  todos in this component.
 */
function TodoList() {
  const [todos, setTodos] = useState([]);

  // Function that will be passed into NewTodoForm to update the todos state
  const addTodo = (todo) => {
    let newTodo = {
      ...todo,
      id: uuid(),
      text: todo.text
    };
    setTodos(todos => [...todos, newTodo]);
  }

  // Function that will be passed into Todo to delete Todo and update the todos state
  const removeTodo = (deleteId) => {
    setTodos(todos => {
      let oldTodos = [...todos];
      const newTodos = oldTodos.filter(targetTodo => 
        deleteId !== targetTodo.id
      );
      return newTodos;
    })
  }

  // Function that will be update List showing all todos
  const renderBoxes = () => {
    return (
      <div className="todolist">
      {todos.map(todo => (
        // <div key={todo.id}> // this would have worked, but no reason to, you can have it on Todo component
        // key has to go on what i map through
          <Todo key={todo.id} id={todo.id} text={todo.text} removeTodo={removeTodo}/>
        // </div>
      ))}
      </div>
    )
  }

  return (
    <div className="todoApp">
      <h2>Todo App</h2>
      {<NewTodoForm addTodo={addTodo} />}
      {renderBoxes()}
    </div>
  )
}

export default TodoList;