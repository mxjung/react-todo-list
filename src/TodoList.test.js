import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
})

it("Adds Todos", function () {
  const { queryByText, getByLabelText, queryAllByText } = render(<TodoList />);

  // fill out the form (for Todo 1)
  const textInput = getByLabelText("Todo:");
  fireEvent.change(textInput, { target: { value: "hello world" } });

  // Click Add Button (1 Box)
  fireEvent.click(queryByText("Add a new Todo!"));
  expect(queryByText("hello world")).toBeInTheDocument();
  expect(queryAllByText("X").length).toBe(1);

  // fill out the form (for Todo 2)
  const textInput2 = getByLabelText("Todo:");
  fireEvent.change(textInput2, { target: { value: "hello world again" } });

  // Click Add Button (2 Boxes)
  fireEvent.click(queryByText("Add a new Todo!"));
  expect(queryByText("hello world again")).toBeInTheDocument();
  expect(queryAllByText("X").length).toBe(2);
})

it("Deletes Todos", function () {
  const { queryByText, getByLabelText, queryAllByText } = render(<TodoList />);

  // fill out the form (for Todo 1)
  const textInput = getByLabelText("Todo:");
  fireEvent.change(textInput, { target: { value: "hello world" } });

  // Click Add Button (1 Box)
  fireEvent.click(queryByText("Add a new Todo!"));
  expect(queryByText("hello world")).toBeInTheDocument();
  expect(queryAllByText("X").length).toBe(1);

  // Click Delete 
  const deleteBtn = queryByText('X');
  fireEvent.click(deleteBtn);

  // Check that Todo element is no longer on page
  expect(queryByText("hello world")).not.toBeInTheDocument();
})