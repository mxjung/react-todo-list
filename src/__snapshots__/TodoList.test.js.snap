// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="todoApp"
  >
    <h2>
      Todo App
    </h2>
    <form>
      <label
        for="text"
      >
        Todo:
      </label>
      <input
        id="text"
        name="text"
        value=""
      />
      <button>
        Add a new Todo!
      </button>
    </form>
    <div
      class="todolist"
    />
  </div>
</DocumentFragment>
`;
